### ISSUE TITLE
objective function

### ISSUE THREAD

**Issue by amosborne** [patrickthomashansen/dorfromantik_helper#3]
Just dropping by to say kudos to you for making this tool. I was thinking about making a similar tool and ended up finding yours... Cool! I was curious how you decided on the objective/evaluation function you are using:

https://github.com/patrickthomashansen/dorfromantik_helper/blob/39274f0872910d637adf25e13ef735f7aca6aaaa/evaluator.py#L47-L54

I was thinking that the best approach would be simply to maximize the number of perfect or potentially-perfect tiles on the board (and therefore only ruin tiles as a very last resort). As a tie-breaker, you would then count how many edges are connected. Further tie-breakers would be up to human judgement in lieu of a more complicated solver.

Your evaluation function sometimes suggests ruining a tile in order to perfect another. I was just curious if that was intentional. The overall result speaks for itself obviously; the tool works quite well.

**Comment by patrickthomashansen** [comment]
Hi Alex,

I'm glad you've found my tool useful! You are correct in that I try to
maximize the number of good connections while minimizing the number of
incompatible connections. Sometimes the weights of the evaluation function
result in ruining a tile in order to get several perfects, and this is
intentional. For example, a location with 5 surrounding tiles would be
quite difficult to find a perfectly fitting tile compared to a location
with 3 surrounding tiles, so sometimes I would prefer to place a tile in
the former location even if it is not perfect, especially early on in a
game when you are working with a smaller deck of tiles and perfects are
more crucial.

I ended up with my evaluation function just by trial and error based on my
own strategy as I play. I weigh good connections that result in a perfect
higher than I do a good connection that does not. Additionally, I penalize
greater for an incompatible connection that causes a tile to no longer be
able to be perfect more than I do an incompatible connection with an
already ruined tile. I could have come up with something more intricate,
but this approach seemed to yield good results.

When I play Dorfromantik, I try to use this tool merely as a guide rather
than an oracle solution, and thus I use it mainly to highlight potential
placement locations. In the case of a tie or close score (or if it helps
with a quest), I will inevitably need to use my own judgement while
playing, which I prefer over having a perfect tool that plays the game for
me :)

I would encourage you to fork and modify my tool if you have a more
sophisticated approach in mind. I would be curious to see what others can
come up with!

Also, I am curious how you found this repo? I made a reddit post a while ago highlighting the repo. I'm surprised people are still using this. I'd be happy to work on the tool more if I knew people were using it.
